extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -394,92 +405,105 @@ export class Bind {
+      : this.initializePromise_;
+
+    return waitFor.then(() =>
+      this.timer_.timeoutPromise(
+        options.timeout || 2000,
+        this.rescan_(addedElements, removedElements, options),
+        'Timed out waiting for amp-bind to rescan.'
+      )
+    );
+  }
+
+  /**
+   * @param {!Array<!Element>} addedElements
+   * @param {!Array<!Element>} removedElements
+   * @param {!BindRescanOptionsDef} options
    * @return {!Promise}
+   * @private
    */
-  scanAndApply(addedElements, removedElements, opt_timeout) {
-    // TODO(choumx): Dependency on init may be removed if we skip elements
-    // with .i-amphtml-binding during tree walk and extract macro setup.
-    return this.initializePromise_.then(() => {
-      return this.rescan_(addedElements, removedElements, opt_timeout || 2000);
+  rescan_(addedElements, removedElements, options) {
+    dev().info(TAG, 'rescan: ', addedElements, removedElements, options);
+
+    const rescanPromise = options.fast
+      ? this.fastScan_(addedElements, removedElements)
+      : this.slowScan_(addedElements, removedElements);
+
+    return rescanPromise.then(() => {
+      if (options.update) {
+        return this.evaluate_().then(results =>
+          this.applyElements_(results, addedElements)
+        );
+      }
     });
   }
 
   /**
    * @param {!Array<!Element>} addedElements
    * @param {!Array<!Element>} removedElements
-   * @param {number} timeout
    * @return {!Promise}
    * @private
    */
-  rescan_(addedElements, removedElements, timeout) {
-    dev().info(TAG, 'rescan:', addedElements, removedElements);
-    /**
-     * Helper function for cleaning up bindings in removed elements.
-     * @param {number} added
-     * @return {!Promise}
-     */
-    const cleanup = added => {
-      this.removeBindingsForNodes_(removedElements).then(removed => {
-        dev().info(
-          TAG,
-          '⤷',
-          'Δ:',
-          added - removed,
-          ', ∑:',
-          this.numberOfBindings()
-        );
-      });
-      return Promise.resolve();
-    };
-    // Early-out if max number of bindings has already been exceeded.
-    // This can happen since we purge old bindings in `removedElements`
-    // _after_ adding new ones (rather than before) as an optimization.
-    if (this.numberOfBindings() > this.maxNumberOfBindings_) {
-      this.emitMaxBindingsExceededError_();
-      return cleanup(0);
-    }
+  fastScan_(addedElements, removedElements) {
+    // Sync remove bindings from internal state first, but don't chain on
+    // returned promise (worker message) as an optimization.
+    const removePromise = this.removeBindingsForNodes_(removedElements);
+
+    // Scan `addedElements` and descendants for bindings.
     const bindings = [];
-    // Scan `addedElements` and their children for elements with bindings.
-    const elementsToScan = addedElements.slice();
+    const elementsToScan = addedElements.filter(el =>
+      el.hasAttribute('i-amphtml-binding')
```

Comment:
Why are you filtering here?

Reply:
Minor optimization. We only want to scan elements with `[i-amphtml-binding]` attribute, and elements in `addedElements` may not have it. 

`scanElement_()` iterates over an element's attributes, so this should be a small perf win.